### ISSUE TITLE
`npm ls` returns an exit code of 1 if it finds unmet peer dependencies

### ISSUE THREAD

**Issue by sonicdoe** [npm/npm#17624]
#### I'm opening this issue because:

npm is doing something I don't understand.

#### What's going wrong?

`npm ls` exits with an exit code of `1` if it finds unmet peer dependencies.

See https://github.com/travis-ci/travis-build/pull/1096#issuecomment-313121633. Both @BanzaiMan and I were surprised that `npm ls` “fails” if there are unmet peer dependencies. While this behavior might be intended, I think it is worth discussing since the actual task of the command (which would be listing dependencies) didn’t really fail.

#### How can the CLI team reproduce the problem?

```sh
$ npm install react-dom
$ npm ls
$ echo $?
```

### supporting information:

 - `npm -v` prints: `5.0.4`
 - `node -v` prints: `v8.1.3`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: macOS

**Comment by zkat** [comment]
`npm ls` is intended for programmatic use and will error if it detects anything wrong with the tree, such as a peerDependency being missing.

**Comment by BanzaiMan** [comment]
If checking the fitness of the dependency tree is the *purpose* of this command, then `ls` (a.k.a. `list`) is a gross misnomer.

**Comment by timoxley** [comment]
@zkat note the exit code behaviour isn't documented in `npm help ls` AFAICT

**Comment by tswaters** [comment]
is it possible to list out the package contents without performing validation and returning non-0 exit code?

this is useful for ci environments to (1)  print out what was installed and (2) not fail the build with a non-0 exit code.

**Comment by ericprud** [comment]
> is it possible to list out the package contents without performing validation and returning non-0 exit code?

Here are two ways you can do this in your `.travis.yml`:

```
before_script:
  - npm list || true
  - find node_modules/ -type f -name package.json -exec echo -n {} \; -exec grep '"version":' {} \;

```

I'm currently using this to track down why npm v5.0.0 is installing webpack@3.10.0 with many unmet dependencies.

**Comment by anwalkers** [comment]
I am also having issues when a package that has a peer dependency with an '||' version definition...

From karma-typescript's [package.json](https://github.com/monounity/karma-typescript/blob/2088ce3dfe0a001fa208c5d534a54e475d23c8de/package.json#L138)

Since we are installing version 2.x of karma we have an unmet dependency for karma@1

```
npm ERR! peer dep missing: karma@1, required by karma-typescript@3.0.9
```

This causes our vulnerability scanner [whitesource](https://github.com/whitesource/npm-plugin) to fail our build pipeline. It uses ```npm ls``` internally to check all of the referenced dependencies.

```
PS > node --version
v8.9.4
PS > npm --version
5.6.0
```

**Comment by unional** [comment]
This is causing builds on travis to fail: https://github.com/travis-ci/travis-ci/issues/9357

**Comment by ripun** [comment]
Please advise solution for below issue
yes when i do npm ls , it fails with eror code 1

 npm ls

├── UNMET DEPENDENCY babel-polyfill@^6.26.0
├── UNMET DEPENDENCY d3@^4.10.0
├── UNMET DEPENDENCY es6-promise@^4.1.1
├── UNMET DEPENDENCY i18next@^8.4.3

npm ERR! missing: babel-polyfill@^6.26.0, required by xxx

**Comment by legodude17** [comment]
That is intentional. If it is a problem you could run `npm ls || true` to swallow the exit code.

**Comment by macroorganizm** [comment]
where i should add this `npm ls || true`?
I tried add it in before_script and it just cause of additional execution of `npm ls` check
Maybe something wrong with my .yaml file?

sudo: required
language: node_js
node_js: 8.8.1
notifications:
  email: false

services:
  - docker

stages:
- name: Precache
- name: Build Front
- name: Build Backend
... another stages ...

cache:
  directories:
    - node_modules
before_install:
- npm install -g npm@5.6.0
before_script:
- npm ls || true
script:
- echo "skipping tests"
jobs:
  fast_finish: true
  include:
  - stage: Precache
    script: true
    install: npm install
  - stage: Build Front
    skip_cleanup: true
    script: npm run build
  - stage: Build Backend
    skip_cleanup: true
    script: npm run server:build
... another jobs...

**Comment by sonicdoe** [comment]
To clarify, it should not be necessary to add `npm ls || true` (or `npm ls`, for that matter) to your `.travis.yml`. By default, Travis CI already executes `npm ls` after installing your dependencies and will continue with your build should `npm ls` exit with a non-zero status code.

@macroorganizm Are you running into any issues when removing the `before_script` part of your `.travis.yml`?

**Comment by macroorganizm** [comment]
yes, i have added it for solving issue, mentioned in the topic head. Occasionally(!) Travis exits with an exit code of 1 if `npm ls` finds unmet peer dependencies. Sometimes it start to happens after few days after changing in package.json. I mean `npm ls` shows errors, but Travis builds still pass during next 1-2 days(last time), or even longer(previously), but later, Travis anyway starts to exit after failing a `npm ls` check. I know it sounds strange(magic!), so i try to investigate what exactly happens

**Comment by sonicdoe** [comment]
@macroorganizm Sounds strange indeed. It might be worth opening an issue in [travis-ci/travis-ci](https://github.com/travis-ci/travis-ci) with links to your Travis CI builds.

**Comment by macroorganizm** [comment]
sadly, it happens in private repo and it's builds is hidden, but thank you, i'll try reproduce it in public.
